cookbooks/retention.md

```diff
@@ -0,0 +1,214 @@
+        if diff >= 7: # exclude first 7 days
+            for period in sorted(PERIODS):
+                if diff <= PERIODS[period]['end']:
+                    return period
+
+@udf(returnType=st.StringType())
+def from_unixtime_handler(ut):
+    """
+    Converts unix time (in days) to a string in %Y%m%d format.
+    This is a spark UDF.
+    
+    Params:
+    ut (int): unix time in days
+    
+    Returns a date as a string if it is parsable by datetime, otherwise None
+    """
+    if ut is not None:
+        try:
+            return (dt.datetime.fromtimestamp(ut * 24 * 60 * 60).strftime("%Y%m%d"))
+        except:
+            return None
+        
+```
+
+Now we can load in a subset of `main_summary` and construct the necessary fields for retention calculations:
+
+```python
+ms = spark.sql("""
+    SELECT 
+        client_id, 
+        submission_date_s3,
+        profile_creation_date,
+        os
+    FROM main_summary 
+    WHERE 
+        submission_date_s3 >= '20180401'
+        AND submission_date_s3 <= '20180603'
+        AND sample_id = '42'
+        AND app_name = 'Firefox'
+        AND normalized_channel = 'release'
+        AND os in ('Darwin', 'Windows_NT', 'Linux')
+    """)
+
+PCD_CUTS = ('20180401', '20180415')
+
+ms = (
+    ms.withColumn("pcd", from_unixtime_handler("profile_creation_date")) # i.e. 17500 -> '20171130'
+      .filter("pcd >= '{}'".format(PCD_CUTS[0]))
+      .filter("pcd <= '{}'".format(PCD_CUTS[1]))
+      .withColumn("period", get_period("pcd", "submission_date_s3"))
+)
+```
+
+Note that we filter to profiles that were created in the first half of April so that we have sufficient time to observe 6 weeks of behavior. Now we can calculate retention!
+
+```python
+os_counts = (
+    ms
+    .groupby("os")
+    .agg(F.countDistinct("client_id").alias("total_clients"))
+)
+
+weekly_counts = (
+    ms
+    .groupby("period", "os")
+    .agg(F.countDistinct("client_id").alias("n_week_clients"))
+)
+
+retention_by_os = (
+    weekly_counts
+    .join(os_counts, on='os')
+    .withColumn("retention", F.col("n_clients") / F.col("total_count"))
+)
+```
+
+Peeking at 6-Week Retention
+
+```python
+retention_by_os.filter("period = 6").show()
+```
+
+```
++----------+------+--------------+-------------+-------------------+
+|        os|period|n_week_clients|total_clients|          retention|
++----------+------+--------------+-------------+-------------------+
+|     Linux|     6|          1495|        22422|0.06667558647756668|
+|    Darwin|     6|          1288|         4734|0.27207435572454586|
+|Windows_NT|     6|         29024|       124872|0.23243000832852842|
++----------+------+--------------+-------------+-------------------+
+
+
+```
+
+we observe that 35.6% of Linux users whose profile was created in the first half of April submitted a ping 6 weeks later, and so forth. The example code snippets are consolidated in [this notebook](https://gist.github.com/benmiroglio/fc708e5905fad33b43adb9c90e38ebf4).
+
+
```

Comment:
A couple of cleaning caveats that might be worth mentioning.

First is the client [submission latency](https://docs.telemetry.mozilla.org/datasets/mozetl/churn/reference.html#background-and-caveats), which can be between somewhere between 2-5 days. `clients_daily` accounts for this when aggregating clients per day, but it's not without it's own issues.

There are also noisy clients with inconsistent profile creation and sub-session dates. Different thresholds will affect the consistency of results. The cookbook could point to an existing baseline in the future.

Reply:
While I agree these are potential issues, I think it is beyond the scope of the cookbook since are general data issues subject to most Firefox analysis. I don't want to bog things down with all potential caveats, but rather serve as a jumping off point methodologically-speaking.

Since retention is usually calculated to compare two or more groups of users, I feel safe to exclude these caveats since each group is subject to the same pitfalls (in this case). 

@acmiyaguchi Let me know if you feel this is justified!

Reply:
Yep, sounds good to me. A six week waiting period and relatively consistent results will probably take care of those caveats across analyses.